ddtrace/bootstrap/sitecustomize.py

```diff
@@ -2,30 +2,108 @@
+        return True
+
+
+else:
+    import importlib
+
+    def is_installed(module_name):
+        return importlib.util.find_spec(module_name)
+
+
+def should_cleanup_loaded_modules():
+    dd_unload_sitecustomize_modules = os.getenv("DD_UNLOAD_MODULES_FROM_SITECUSTOMIZE", default="0").lower()
+    if dd_unload_sitecustomize_modules not in ("1", "auto"):
+        return False
+    elif dd_unload_sitecustomize_modules == "auto" and not any(
+        is_installed(module_name) for module_name in MODULES_THAT_TRIGGER_CLEANUP_WHEN_INSTALLED
+    ):
+        return False
+    return True
+
+
+def cleanup_loaded_modules(aggressive=False):
+    """
+    "Aggressive" here means "cleanup absolutely every module that has been loaded since startup".
+    Non-aggressive cleanup entails leaving untouched certain modules
+    This distinction is necessary because this function is used both to prepare for gevent monkeypatching
+    (requiring aggressive cleanup) and to implement "module cloning" (requiring non-aggressive cleanup)
+    """
+    # Figuring out modules_loaded_since_startup is necessary because sys.modules has more in it than just what's in
+    # import statements in this file, and unloading some of them can break the interpreter.
+    modules_loaded_since_startup = set(_ for _ in sys.modules if _ not in MODULES_LOADED_AT_STARTUP)
+    # Unload all the modules that we have imported, except for ddtrace and a few
+    # others that don't like being cloned.
+    # Doing so will allow ddtrace to continue using its local references to modules unpatched by
+    # gevent, while avoiding conflicts with user-application code potentially running
+    # `gevent.monkey.patch_all()` and thus gevent-patched versions of the same modules.
+    for module_name in modules_loaded_since_startup:
+        if aggressive:
+            del sys.modules[module_name]
+            continue
+
+        for module_to_not_cleanup in MODULES_TO_NOT_CLEANUP:
+            if module_name == module_to_not_cleanup:
+                break
+            elif module_name.startswith("%s." % module_to_not_cleanup):
+                break
+        else:
+            del sys.modules[module_name]
+
+
+will_run_module_cloning = should_cleanup_loaded_modules()
+if not will_run_module_cloning:
+    # Perform gevent patching as early as possible in the application before
+    # importing more of the library internals.
+    if os.environ.get("DD_GEVENT_PATCH_ALL", "false").lower() in ("true", "1"):
+        # successfully running `gevent.monkey.patch_all()` this late into
+        # sitecustomize requires aggressive module unloading beforehand.
+        # gevent's documentation strongly warns against calling monkey.patch_all() anywhere other
+        # than the first line of the program. since that's what we're doing here,
+        # we cleanup aggressively beforehand to replicate the conditions at program start
+        # as closely as possible.
+        cleanup_loaded_modules(aggressive=True)
```

Comment:
I suppose we can be aggressive here because we still haven't loaded much?

Reply:
Yes, that's probably true.